core/src/main/java/io/methvin/watcher/PathUtils.java

```diff
@@ -42,17 +43,17 @@ public static <T> SortedMap<Path, T> subMap(SortedMap<Path, T> pathMap, Path tre
     return pathMap.subMap(treeRoot, Paths.get(treeRoot.toString(), "" + Character.MAX_VALUE));
   }
 
-  public static SortedMap<Path, FileHash> createHashCodeMap(Path file, FileHasher fileHasher)
-      throws IOException {
-    return createHashCodeMap(Collections.singletonList(file), fileHasher);
+  public static SortedMap<Path, FileHash> createHashCodeMap(
+      Path file, FileHasher fileHasher, FileTreeVisitor fileTreeVisitor) throws IOException {
+    return createHashCodeMap(Collections.singletonList(file), fileHasher, fileTreeVisitor);
   }
 
-  public static SortedMap<Path, FileHash> createHashCodeMap(List<Path> files, FileHasher fileHasher)
-      throws IOException {
+  public static SortedMap<Path, FileHash> createHashCodeMap(
+      List<Path> files, FileHasher fileHasher, FileTreeVisitor fileTreeVisitor) throws IOException {
     SortedMap<Path, FileHash> lastModifiedMap = new ConcurrentSkipListMap<>();
     if (fileHasher != null) {
       for (Path file : files) {
-        for (Path child : recursiveListFiles(file)) {
+        for (Path child : recursiveListFiles(fileTreeVisitor, file)) {
           FileHash hash = PathUtils.hash(fileHasher, child);
```

Comment:
So I think now errors are handled differently in `createHashCodeMap` than in `initWatcherState`. Since `initWatcherState` calls `PathUtils.hash` inside the `Callback`, an exception during hashing would be caught by `onFailure`. In `createHashCodeMap` we're using the visitor to just get a list of the files and doing the hashing outside the callback, so an exception would not be caught.

The main difference here is that `createHashCodeMap` is used by `MacOSXListeningWatchService` whereas `initWatcherState` is used by `DirectoryWatcher` directly, but they basically do the same thing so it probably makes sense to have consistent behavior.